sys/include/atomic_utils.h

```diff
@@ -0,0 +1,1053 @@
+/*
+ * Copyright (C) 2020 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @defgroup    sys_atomic_utils    Utility functions for atomic access
+ * @ingroup     sys
+ *
+ * This modules adds some utility functions to perform atomic accesses.
+ *
+ * # Usage
+ *
+ * The atomic utilitys allow atomic access to regular variables.
+ *
+ *  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
+ *  uint32_t global_counter;
+ *
+ *  void irq_handler(void)
+ *  {
+ *      // No need to use atomic access in IRQ handlers: RIOT does not
```

Comment:
Isn't this a classic example of where compilers can optimize in unexpected ways, as soon as more code is added?

E.g.,

```
uint32_t global_counter;
 void irq_handler(void)
 {
   while (condition) {
     // do_register_magic
   }
   global_counter++;
}
```

The compiler is allowed to write to global_counter whenever it thinks is best.

Reply:
What I mean is, using atomic would ensure the compiler doesn't optimize around that write, correct?

(I know I argued before that *our* compilers on *our* architectures don't do those reorders... :innocent: )

Reply:
We have (hopefully :-D) barriers in place at the beginning and end of each ISR. That way the compiler can reorder the writes within the ISR, but they should happen somewhere within the ISR. Also other optimizations like elimination of dead writes etc. can work properly, but the very last write cannot be assumed to be dead due to those barriers.

Reply:
right, I got confused with doing this in threads. nevermind